Audiobook Player/Models/Book.swift

```diff
@@ -0,0 +1,148 @@
+//
+//  Book.swift
+//  Audiobook Player
+//
+//  Created by Florian Pichler on 09.04.18.
+//  Copyright © 2018 Tortuga Power. All rights reserved.
+//
+
+import UIKit
+import AVFoundation
+
+class Book: NSObject {
+    // MARK: Meta data
+
+    var artwork: UIImage
+    var asset: AVAsset
+    var fileURL: URL
+    var title: String = ""
+    var author: String = ""
+
+    var identifier: String {
+        return self.fileURL.lastPathComponent
+    }
+
+    var displayTitle: String {
+        return title + " - " + author
+    }
+
+    // MARK: Time
+
+    var duration: TimeInterval {
+        return TimeInterval(CMTimeGetSeconds(self.asset.duration))
+    }
+
+    var currentTime: TimeInterval = 0.0 {
+        didSet {
+            // Store state only every 2 seconds, I/O can be expensive
+            if Int(self.currentTime) % 2 == 0 && self.currentTime != oldValue {
+                UserDefaults.standard.set(self.currentTime, forKey: self.identifier)
+            }
+
+            self.updateCurrentChapter()
+            self.updatePercentCompleted()
+        }
+    }
+
+    private func getStoredTime() -> TimeInterval {
+        // get stored value for current time of book in seconds
+        var lastPlayedPosition = UserDefaults.standard.double(forKey: self.identifier)
+
+        // If smartRewind is enabled and time since last play was 10 minutes (600s), rewind audiobook by 30 seconds or to start.
+        if
+            let lastPlayTime = UserDefaults.standard.object(forKey: UserDefaultsConstants.lastPauseTime+"_\(self.identifier)") as? Date,
+            UserDefaults.standard.bool(forKey: UserDefaultsConstants.smartRewindEnabled)
+        {
+            if Date().timeIntervalSince(lastPlayTime) >= 600.0 {
+                lastPlayedPosition = max(lastPlayedPosition - 30.0, 0)
+
+                UserDefaults.standard.set(nil, forKey: UserDefaultsConstants.lastPauseTime+"_\(self.identifier)")
+            }
+        }
+
+        return lastPlayedPosition
+    }
+
+    var percentCompleted: Int = 0
+    var percentCompletedString: String = "0%"
```

Comment:
we could have this one be a computed property too based on `percentCompleted`

Reply:
I didn't specify, but I'd rather make `percentCompletedString` a `get` only variable, that way we handle one less state in this class

Reply:
That `get` would be based on `percentCompleted`